lib/trie/database.go

```diff
@@ -61,241 +61,270 @@ func (t *Trie) store(db chaindb.Batch, curr Node) error {
+	branch, ok := n.(*node.Branch)
 	if !ok {
 		return
 	}
 
-	for i, child := range c.Children {
+	for i, child := range branch.Children {
 		if child == nil {
 			continue
 		}
 
-		proofNode, ok := proof[common.BytesToHex(child.GetHash())]
+		proofHash := common.BytesToHex(child.GetHash())
+		node, ok := proofHashToNode[proofHash]
 		if !ok {
 			continue
 		}
+		delete(proofHashToNode, proofHash)
 
-		c.Children[i] = proofNode
-		t.loadProof(proof, proofNode)
+		branch.Children[i] = node
+		t.loadProof(proofHashToNode, node)
 	}
 }
 
 // Load reconstructs the trie from the database from the given root hash.
 // It is used when restarting the node to load the current state trie.
-func (t *Trie) Load(db chaindb.Database, root common.Hash) error {
-	if root == EmptyHash {
+func (t *Trie) Load(db chaindb.Database, rootHash common.Hash) error {
+	if rootHash == EmptyHash {
 		t.root = nil
 		return nil
 	}
 
-	enc, err := db.Get(root[:])
+	rootHashBytes := rootHash[:]
+
+	encodedNode, err := db.Get(rootHashBytes)
 	if err != nil {
-		return fmt.Errorf("failed to find root key=%s: %w", root, err)
+		return fmt.Errorf("failed to find root key %s: %w", rootHash, err)
 	}
 
-	t.root, err = node.Decode(bytes.NewReader(enc))
+	reader := bytes.NewReader(encodedNode)
+	root, err := node.Decode(reader)
 	if err != nil {
-		return err
+		return fmt.Errorf("cannot decode root node: %w", err)
 	}
-
+	t.root = root
 	t.root.SetDirty(false)
-	t.root.SetEncodingAndHash(enc, root[:])
+	t.root.SetEncodingAndHash(encodedNode, rootHashBytes)
 
 	return t.load(db, t.root)
 }
 
-func (t *Trie) load(db chaindb.Database, curr Node) error {
-	if c, ok := curr.(*node.Branch); ok {
-		for i, child := range c.Children {
-			if child == nil {
-				continue
-			}
+func (t *Trie) load(db chaindb.Database, n Node) error {
+	branch, ok := n.(*node.Branch)
+	if !ok {
+		return nil
+	}
 
-			hash := child.GetHash()
-			enc, err := db.Get(hash)
-			if err != nil {
-				return fmt.Errorf("failed to find node key=%x index=%d: %w", hash, i, err)
-			}
+	for i, child := range branch.Children {
+		if child == nil {
+			continue
+		}
 
-			child, err = node.Decode(bytes.NewReader(enc))
-			if err != nil {
-				return err
-			}
+		hash := child.GetHash()
+		encodedNode, err := db.Get(hash)
+		if err != nil {
+			return fmt.Errorf("cannot find child node key 0x%x in database: %w", hash, err)
+		}
 
-			child.SetDirty(false)
-			child.SetEncodingAndHash(enc, hash)
+		reader := bytes.NewReader(encodedNode)
+		decodedNode, err := node.Decode(reader)
+		if err != nil {
+			return fmt.Errorf("cannot decode node with hash 0x%x: %w", hash, err)
+		}
 
-			c.Children[i] = child
-			err = t.load(db, child)
-			if err != nil {
-				return err
-			}
+		decodedNode.SetDirty(false)
+		decodedNode.SetEncodingAndHash(encodedNode, hash)
+		branch.Children[i] = decodedNode
+
+		err = t.load(db, decodedNode)
+		if err != nil {
+			return fmt.Errorf("cannot load child with hash 0x%x: %w", hash, err)
```

Comment:
Would be nice to log the index here
```suggestion
			return fmt.Errorf("cannot load child at index %d with hash 0x%x: %w", i, hash, err)
```